src/main/java/org/opensearch/flowframework/template/TemplateParser.java

```diff
@@ -0,0 +1,164 @@
+                    }
+
+                    @Override
+                    public Map<String, String> getParams() {
+                        // See RestCreateIndexAction for source of param keys needed
+                        return Map.of("index", request.index());
+                    }
+
+                };
+            }
+            nodes.add(new ProcessNode(nodeId, workflowStep, inputData));
+        }
+
+        for (JsonElement edgeJson : graph.getAsJsonArray(EDGES)) {
+            JsonObject edgeObject = edgeJson.getAsJsonObject();
+            String sourceNodeId = edgeObject.get(SOURCE).getAsString();
+            String destNodeId = edgeObject.get(DESTINATION).getAsString();
+            if (sourceNodeId.equals(destNodeId)) {
+                throw new IllegalArgumentException("Edge connects node " + sourceNodeId + " to itself.");
+            }
+            edges.add(new ProcessSequenceEdge(sourceNodeId, destNodeId));
+        }
+
+        return topologicalSort(nodes, edges);
+    }
+
+    private static List<ProcessNode> topologicalSort(List<ProcessNode> nodes, List<ProcessSequenceEdge> edges) {
+        // Define the graph
+        Set<ProcessSequenceEdge> graph = new HashSet<>(edges);
+        // Map node id string to object
+        Map<String, ProcessNode> nodeMap = nodes.stream().collect(Collectors.toMap(ProcessNode::id, Function.identity()));
+        // Build predecessor and successor maps
+        Map<ProcessNode, Set<ProcessSequenceEdge>> predecessorEdges = new HashMap<>();
+        Map<ProcessNode, Set<ProcessSequenceEdge>> successorEdges = new HashMap<>();
+        for (ProcessSequenceEdge edge : edges) {
+            ProcessNode source = nodeMap.get(edge.getSource());
+            ProcessNode dest = nodeMap.get(edge.getDestination());
+            predecessorEdges.computeIfAbsent(dest, k -> new HashSet<>()).add(edge);
+            successorEdges.computeIfAbsent(source, k -> new HashSet<>()).add(edge);
+        }
+        // update predecessors on the node object
+        nodes.stream().filter(n -> predecessorEdges.containsKey(n)).forEach(n -> {
+            n.setPredecessors(predecessorEdges.get(n).stream().map(e -> nodeMap.get(e.getSource())).collect(Collectors.toSet()));
+        });
+
+        // See https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
+        // L <- Empty list that will contain the sorted elements
+        List<ProcessNode> sortedNodes = new ArrayList<>();
+        // S <- Set of all nodes with no incoming edge
+        Queue<ProcessNode> sourceNodes = new ArrayDeque<>();
+        nodes.stream().filter(n -> !predecessorEdges.containsKey(n)).forEach(n -> sourceNodes.add(n));
+        if (sourceNodes.isEmpty()) {
+            throw new IllegalArgumentException("No start node detected: all nodes have a predecessor.");
+        }
+        logger.debug("Start node(s): {}", sourceNodes);
```

Comment:
Don't we want to log just the first node here and not an entire queue?

Reply:
We do log the entire queue later.  Sample output in the PR:
```
17:22:34.597 INFO  o.o.f.demo.Demo - Parsing graph to sequence...
17:22:34.634 DEBUG o.o.f.template.TemplateParser - Start node(s): [fetch_model]
17:22:34.635 DEBUG o.o.f.template.TemplateParser - Execution sequence: [fetch_model, create_search_pipeline, create_ingest_pipeline, create_neural_search_index]
```